### PR TITLE
add data bundles to sdk

### PR DESCRIPTION
**Pull request by rohandandavati**

### CODE REVIEW

File: fiamui-app/src/main/java/com/example/firebase/fiamui/MainActivity.java [hunk 1]
```diff
@@ -154,13 +156,19 @@ public void onStartClick(View v) {
     String secondaryActionButtonTextString = secondaryActionButtonText.getText().toString();
 
     CampaignMetadata campaignMetadata = new CampaignMetadata("test_campaign", "name", true);
+    Map<String, String> data =
+        new HashMap<String, String>() {
+          {
+            put("up", "dog");
```
Comment (prakhar1989) [inline fiamui-app/src/main/java/com/example/firebase/fiamui/MainActivity.java]: Lets use `com.google.common.collect.ImmutableMap` for the map

File: firebase-inappmessaging/src/test/java/com/google/firebase/inappmessaging/internal/AnalyticsEventsManagerTest.java [hunk 2]
```diff
@@ -60,9 +60,9 @@
   private static final TriggeringCondition.Builder onAnalyticsEvent =
       TriggeringCondition.newBuilder().setEvent(Event.newBuilder().setName(ANALYTICS_EVENT_1));
   private static final TriggeringCondition onForeground =
-      TriggeringCondition.newBuilder().setFiamTrigger(ON_FOREGROUND).build();
+      (TriggeringCondition) TriggeringCondition.newBuilder().setFiamTrigger(ON_FOREGROUND).build();
```
Comment (prakhar1989) [inline firebase-inappmessaging/src/test/java/com/google/firebase/inappmessaging/internal/AnalyticsEventsManagerTest.java]: Why are these casts required? (here and on `Priority`)

File: firebase-inappmessaging/src/test/java/com/google/firebase/inappmessaging/internal/AnalyticsEventsManagerTest.java [hunk 3]
```diff
@@ -112,11 +112,12 @@ public void constuctor_ok() throws Exception {
   @Test
   public void extractAnalyticsEventNames_filtersOutFiamEvents() {
     FetchEligibleCampaignsResponse campaignsResponse =
-        FetchEligibleCampaignsResponse.newBuilder()
-            .setExpirationEpochTimestampMillis(FUTURE)
-            .addMessages(ANALYTICS_EVENT_THICK_CONTENT_BUILDER)
-            .addMessages(FOREGROUND_THICK_CONTENT_BUILDER)
-            .build();
+        (FetchEligibleCampaignsResponse)
```
Comment (prakhar1989) [inline firebase-inappmessaging/src/test/java/com/google/firebase/inappmessaging/internal/AnalyticsEventsManagerTest.java]: Not sure why this is required

File: firebase-inappmessaging/src/test/java/com/google/firebase/inappmessaging/internal/AnalyticsEventsManagerTest.java [hunk 4]
```diff
@@ -137,11 +138,12 @@ public void extractAnalyticsEventNames_addsEventsWhenResponseHasMultipleTriggers
             .setContent(Content.getDefaultInstance());
 
     FetchEligibleCampaignsResponse campaignsResponse =
-        FetchEligibleCampaignsResponse.newBuilder()
-            .setExpirationEpochTimestampMillis(FUTURE)
-            .addMessages(ANALYTICS_EVENT_THICK_CONTENT_BUILDER)
-            .addMessages(event2ContentBuilder)
-            .build();
+        (FetchEligibleCampaignsResponse)
```
Comment (prakhar1989) [inline firebase-inappmessaging/src/test/java/com/google/firebase/inappmessaging/internal/AnalyticsEventsManagerTest.java]: same

File: firebase-inappmessaging/src/testData/inappmessaging/testutil/TestData.java [hunk 5]
```diff
@@ -39,6 +41,10 @@
       new CampaignMetadata(CAMPAIGN_ID_STRING, CAMPAIGN_NAME_STRING, IS_NOT_TEST_MESSAGE);
   public static final CampaignMetadata TEST_CAMPAIGN_METADATA_MODEL =
       new CampaignMetadata(CAMPAIGN_ID_STRING, CAMPAIGN_NAME_STRING, IS_TEST_MESSAGE);
+  public static final Map<String, String> DATA = new HashMap<>();
+  static {
+    DATA.put("up", "dog");
+  }
```
Comment (prakhar1989) [inline firebase-inappmessaging/src/testData/inappmessaging/testutil/TestData.java]: Lets use `com.google.common.collect.ImmutableMap` to create an immutable map here

```java
public static final Map<String, String> DATA = ImmutableMap.of("foo", "bar");
```

File: firebase-inappmessaging/src/test/java/com/google/firebase/inappmessaging/internal/InAppMessageStreamManagerTest.java [hunk 6]
```diff
@@ -107,10 +107,10 @@
...
 
   private static final CampaignImpressionList CAMPAIGN_IMPRESSIONS =
-      CampaignImpressionList.newBuilder()
-          .addAlreadySeenCampaigns(
-              CampaignImpression.newBuilder().setCampaignId(CAMPAIGN_ID_STRING).build())
-          .build();
+          CampaignImpressionList.newBuilder()
+              .addAlreadySeenCampaigns(
+                      CampaignImpression.newBuilder().setCampaignId(CAMPAIGN_ID_STRING).build())
+              .build();
```
Comment (JasonAHeron) [inline firebase-inappmessaging/src/test/java/com/google/firebase/inappmessaging/internal/InAppMessageStreamManagerTest.java]: I wish these lines could not be marked as changed :( everywhere git blame is now going to say "Rohan did it"

File: firebase-inappmessaging/src/test/java/com/google/firebase/inappmessaging/internal/TestDeviceHelperTest.java [hunk 7]
```diff
@@ -36,7 +37,8 @@
   private static final CampaignProto.ThickContent.Builder thickContentBuilder =
       CampaignProto.ThickContent.newBuilder()
           .setContent(MessagesProto.Content.getDefaultInstance());
-  private static final CampaignProto.ThickContent thickContent = thickContentBuilder.build();
+  private static final CampaignProto.ThickContent thickContent =
+      (ThickContent) thickContentBuilder.build();
```
Comment (JasonAHeron) [inline firebase-inappmessaging/src/test/java/com/google/firebase/inappmessaging/internal/TestDeviceHelperTest.java]: remove cast

File: firebase-inappmessaging/src/test/java/com/google/firebase/inappmessaging/model/ProtoMarshallerClientTest.java [hunk 8]
```diff
@@ -345,26 +349,75 @@ public void card_missingBackgroundColor_throwsException() {
...
     MessagesProto.Content minimumCard =
-        MessagesProto.Content.newBuilder()
-            .setCard(
-                MessagesProto.CardMessage.newBuilder()
-                    .setPrimaryActionButton(BUTTON_PROTO)
-                    .setTitle(TITLE_PROTO)
-                    .setBackgroundHexColor(MESSAGE_BACKGROUND_HEX_STRING)
-                    .setPortraitImageUrl(IMAGE_URL_STRING))
-            .build();
+        (Content)
```
Comment (JasonAHeron) [inline firebase-inappmessaging/src/test/java/com/google/firebase/inappmessaging/model/ProtoMarshallerClientTest.java]: remove cast

File: firebase-inappmessaging/firebase-inappmessaging.gradle [hunk 9]
```diff
@@ -139,7 +139,9 @@ dependencies {
     testImplementation "com.google.truth:truth:$googleTruthVersion"
     testImplementation 'junit:junit:4.12'
     testImplementation 'androidx.test:runner:1.2.0'
-    testImplementation 'org.robolectric:robolectric:4.2'
+    testImplementation ('org.robolectric:robolectric:4.2') {
+        exclude group: 'com.google.protobuf', module: 'protobuf-java'
```
Comment (prakhar1989) [inline firebase-inappmessaging/firebase-inappmessaging.gradle]: Why is this change required?

File: firebase-inappmessaging/src/main/java/com/google/firebase/inappmessaging/model/CardMessage.java [hunk 10]
```diff
@@ -239,7 +241,7 @@ public Builder setBody(@Nullable Text body) {
       return this;
     }
 
-    public CardMessage build(CampaignMetadata campaignMetadata) {
+    public CardMessage build(CampaignMetadata campaignMetadata, Map<String, String> data) {
       if (primaryAction == null) {
         throw new IllegalArgumentException("Card model must have a primary action");
```
Comment (prakhar1989) [inline firebase-inappmessaging/src/main/java/com/google/firebase/inappmessaging/model/CardMessage.java]: Can this be `@Nullable`?

File: firebase-inappmessaging/src/main/java/com/google/firebase/inappmessaging/model/BannerMessage.java [hunk 11]
```diff
@@ -171,7 +173,7 @@ public Builder setBackgroundHexColor(@Nullable String backgroundHexColor) {
       return this;
     }
 
-    public BannerMessage build(CampaignMetadata campaignMetadata) {
+    public BannerMessage build(CampaignMetadata campaignMetadata, Map<String, String> data) {
```
Comment (prakhar1989) [inline firebase-inappmessaging/src/main/java/com/google/firebase/inappmessaging/model/BannerMessage.java]: Should this be `@Nullable`?

File: firebase-inappmessaging/src/main/java/com/google/firebase/inappmessaging/model/ModalMessage.java [hunk 12]
```diff
@@ -171,7 +173,7 @@ public Builder setBackgroundHexColor(@Nullable String backgroundHexColor) {
       return this;
     }
 
-    public ModalMessage build(CampaignMetadata campaignMetadata) {
+    public ModalMessage build(CampaignMetadata campaignMetadata, Map<String, String> data) {
```
Comment (prakhar1989) [inline firebase-inappmessaging/src/main/java/com/google/firebase/inappmessaging/model/ModalMessage.java]: Should this be `@Nullable`?

File: firebase-inappmessaging/firebase-inappmessaging.gradle (hunk 9 above)
Comment (rohandandavati) [reply firebase-inappmessaging/firebase-inappmessaging.gradle]: This allows ProtoMarshallerClientTest.java to run - without this, we get makeImmutable() errors - I don't 100% understand why, but I think it has to do with conflicting versions of proto
